src/main/java/net/snowflake/client/core/arrow/fullvectorconverters/TimestampVectorConverter.java

```diff
@@ -0,0 +1,166 @@
+  private RootAllocator allocator;
+  private ValueVector vector;
+  private DataConversionContext context;
+  private TimeZone timeZoneToUse;
+  private boolean isNTZ;
+
+  /** Field names of the struct vectors used by timestamp */
+  private static final String FIELD_NAME_EPOCH = "epoch"; // seconds since epoch
+
+  private static final String FIELD_NAME_TIME_ZONE_INDEX = "timezone"; // time zone index
+  private static final String FIELD_NAME_FRACTION = "fraction"; // fraction in nanoseconds
+
+  public TimestampVectorConverter(
+      RootAllocator allocator,
+      ValueVector vector,
+      DataConversionContext context,
+      TimeZone timeZoneToUse,
+      boolean isNTZ) {
+    this.allocator = allocator;
+    this.vector = vector;
+    this.context = context;
+    this.timeZoneToUse = timeZoneToUse;
+    this.isNTZ = isNTZ;
+  }
+
+  private IntVector makeVectorOfZeroes(int length) {
+    IntVector vector = new IntVector(FIELD_NAME_FRACTION, allocator);
+    vector.allocateNew(length);
+    vector.zeroVector();
+    vector.setValueCount(length);
+    return vector;
+  }
+
+  private IntVector makeVectorOfUTCOffsets(int length) {
+    IntVector vector = new IntVector(FIELD_NAME_TIME_ZONE_INDEX, allocator);
+    vector.allocateNew(length);
+    vector.setValueCount(length);
+    for (int i = 0; i < length; i++) {
+      vector.set(i, 1440);
+    }
+    return vector;
+  }
+
+  private SFPair<BigIntVector, IntVector> normalizeTimeSinceEpoch(BigIntVector vector) {
+    int length = vector.getValueCount();
+    int scale = Integer.parseInt(vector.getField().getMetadata().get("scale"));
+    if (scale == 0) {
+      IntVector fractions = makeVectorOfZeroes(length);
+      fractions
+          .getValidityBuffer()
+          .setBytes(0L, vector.getValidityBuffer(), 0L, fractions.getValidityBuffer().capacity());
+      return SFPair.of(vector, fractions);
+    }
+    long scaleFactor = ArrowResultUtil.powerOfTen(scale);
+    long fractionScaleFactor = ArrowResultUtil.powerOfTen(9 - scale);
+    BigIntVector epoch = new BigIntVector(FIELD_NAME_EPOCH, allocator);
+    epoch.allocateNew(length);
+    epoch.setValueCount(length);
+    IntVector fractions = new IntVector(FIELD_NAME_FRACTION, allocator);
+    fractions.allocateNew(length);
+    fractions.setValueCount(length);
+    for (int i = 0; i < length; i++) {
+      epoch.set(i, vector.get(i) / scaleFactor);
+      fractions.set(i, (int) ((vector.get(i) % scaleFactor) * fractionScaleFactor));
+    }
+    return SFPair.of(epoch, fractions);
+  }
+
+  private IntVector makeTimeZoneOffsets(
+      BigIntVector seconds, IntVector fractions, TimeZone timeZone) {
+    IntVector offsets = new IntVector(FIELD_NAME_TIME_ZONE_INDEX, allocator);
+    offsets.allocateNew(vector.getValueCount());
+    for (int i = 0; i < vector.getValueCount(); i++) {
+      offsets.set(
+          i,
+          1440
+              + timeZone.getOffset(seconds.get(i) * 1000 + fractions.get(i) / 1000000)
```

Comment:
Could we define & use consts for these calculations?

Reply:
Good catch, forgot to do that after initial testing